scripts/joint_goal_publisher.py

```diff
@@ -0,0 +1,338 @@
+                if minval == maxval:  # this is fixed joint
+                    continue
+
+                self.joint_list.append(name)
+                joint = {'min':minval*pi/180.0, 'max':maxval*pi/180.0, 'zero':0, 'position':0, 'velocity':0, 'effort':0}
+                self.free_joints[name] = joint
+
+    def init_urdf(self, robot):
+        """
+        reads the controllable joints from a urdf
+        :param robot:
+        :return:
+        """
+        robot = robot.getElementsByTagName('robot')[0]
+        # Find all non-fixed joints that are controlled by giskard
+        for child in robot.childNodes:
+            if child.nodeType is child.TEXT_NODE:
+                continue
+            if child.localName == 'joint':
+                jtype = child.getAttribute('type')
+                if jtype in ['fixed', 'floating', 'planar']:
+                    continue
+                name = child.getAttribute('name')
+                if name not in self.giskard_joints:
+                    continue
+
+                self.joint_list.append(name)
+                if jtype == 'continuous':
+                    minval = -pi
+                    maxval = pi
+                else:
+                    try:
+                        limit = child.getElementsByTagName('limit')[0]
+                        minval = float(limit.getAttribute('lower'))
+                        maxval = float(limit.getAttribute('upper'))
+                    except:
+                        rospy.logwarn("%s is not fixed, nor continuous, but limits are not specified!" % name)
+                        continue
+
+                safety_tags = child.getElementsByTagName('safety_controller')
+                if self.use_small and len(safety_tags) == 1:
+                    tag = safety_tags[0]
+                    if tag.hasAttribute('soft_lower_limit'):
+                        minval = max(minval, float(tag.getAttribute('soft_lower_limit')))
+                    if tag.hasAttribute('soft_upper_limit'):
+                        maxval = min(maxval, float(tag.getAttribute('soft_upper_limit')))
+
+                mimic_tags = child.getElementsByTagName('mimic')
+                if self.use_mimic and len(mimic_tags) == 1:
+                    tag = mimic_tags[0]
+                    entry = {'parent': tag.getAttribute('joint')}
+                    if tag.hasAttribute('multiplier'):
+                        entry['factor'] = float(tag.getAttribute('multiplier'))
+                    if tag.hasAttribute('offset'):
+                        entry['offset'] = float(tag.getAttribute('offset'))
+
+                    self.dependent_joints[name] = entry
+                    continue
+
+                if name in self.dependent_joints:
+                    continue
+
+                if self.zeros and name in self.zeros:
+                    zeroval = self.zeros[name]
+                elif minval > 0 or maxval < 0:
+                    zeroval = (maxval + minval)/2
```

Comment:
You can do `from __future__ import division` to make sure that results of division are floats, even if you divide ints. Without import `1/2=0`, with import `1/2=0.5`.